### ISSUE TITLE
Report integration with Sphinx

### ISSUE THREAD

**Issue by Paebbels** [coveragepy/coveragepy#1730]
**Is your feature request related to a problem? Please describe.**

Usually, all my Python projects are documented using Sphinx. Such a documentation contains also an auto generated Python Class Reference as well as multiple code quality indications / analysis reports. Currently this is: code coverage by `coveragypy`, unit tests by `pytest`, static typing analysis by `mypy` and documentation coverage by `docstr_coverage`.

Most tools provide a machine readable output format (e.g. in XML or JSON) and almost all are writing or converting them to an HTML report. This can be a single file or multiple files in a directory.

For integration into the overall documentation, multiple CI jobs generate the results and generated HTML pages and then a final CI job combines all artifacts to a combined static website for e.g. GitHub Pages.

In the last days I investigated and prototyped a new Sphinx extension, which integrates a documentation coverage report directly as native Sphinx content into the documentation. This offers some huge benefits:
* reports are part of the overall design / no style break
* reports are also part of other Sphinx backend outputs e.g. PDF documentation
* parts of the report can be referenced by the main documentation
* reports are part of the main navigation / table of content

I would like to discuss if and how to integrate `coveragepy` into such a Sphinx extension. My extension `sphinx-reports` is open to more report formats like code coverage.

This is an example from my prototype for documentation coverage:
![image](https://github.com/nedbat/coveragepy/assets/956109/685b0002-ca49-4f32-b576-db31069f83bd)

**Describe the solution you'd like**
I would like to read and render code coverage results into Sphinx documentation pages.

I would like to address it in 2 steps:
1. Integrate `coveragepy`'s summary page and then link to auto generated coverage pages.  
   (I haven't figured out how to write complex Sphinx content like colored source code.)
2. Also write source code files via Sphinx.

The extensions alpha code is here: https://pytooling.github.io/sphinx-reports

**Describe alternatives you've considered**
I checked, if auto generated HTML code from `coveragepy` could be copied into a placeholder Sphinx page. While content was transferred, I found huge problems with CSS and Javascript contents. It would results in dozens - if not hundreds - of regex-based rewrite and transformation rules. This can't be maintained.

An other alternative would be `coveragepy` offering ReST (Restructured Text) output and such files could be referenced by a `toctree` directive.

**Additional context**
I believe that expressing the quality of an open-source project is a need to convince others to use and maybe invest time in someone open-source project. It also can shorten some discussions if a library can be used easily and safely.

**Comment by Paebbels** [comment]
Where can I find the documentation of `status.json` written into the HTML output directory. It looks like `.... "nums": [0, 1, 246, 0, 105, 86, 3, 33],` contains a summary of total, covered and uncovered elements.

**Comment by Paebbels** [comment]
**Update:** `status.json` doesn't contain the correct (or final) coverage values. I'm now switching to a full coverage file in JSON format.

Source: https://pytooling.github.io/sphinx-reports/coverage/index.html  
![image](https://github.com/nedbat/coveragepy/assets/956109/3535a6ff-3732-444c-b113-6a8f0797ae74)

**Comment by nedbat** [comment]
I think you've already worked this out, but status.json is not meant for external code to consume: it's a caching mechanism to speed up the HTML report.  The `coverage json` command writes out data you can use.

I'm not sure the simplest way to get what you need.  It seem to me like a niche need to have these reports included in Sphinx generated pages, so I want to find a path for you that doesn't change coverage too much.

**Comment by Paebbels** [comment]
As I transitioned to the "bigger" JSON, I'm already in the phase reading and using all coverage data. I hoped to get along with `status.json` as it's quite small compared to a full JSON or XML.

So I think ReST output or so from Coverage.py is not needed.

At next I want to figure out how to create subpages in Sphinx and how to colorize source code. This is currently black magic in Sphinx.

What might be interesting for me is the sorting feature that you use in your HTML tables.

Do you have suggestions on level for the coloring?  
So what's bad, medium, good very good coverage?

**Comment by nedbat** [comment]
Hi, I don't have suggestions for the levels for various colors.  Probably people will want it to be configurable, but that's up to you.  It sounds like you have everything you need from me?

**Comment by Paebbels** [comment]
>  Probably people will want it to be configurable

Yes, the number of levels and the percentage can be configured. I plan to have some kind of default coloring, but otherwise the user colors are used. Actually it's a percentage, description text (good, bad, very bad, ..) and a CSS class name.

>  It sounds like you have everything you need from me?

May be this part:
> What might be interesting for me is the sorting feature that you use in your HTML tables.

The Javascript and column sorting is this code written/owned by you or some kind of library? I think it might be good to have a hierarchy and a sorted view. I see Coverage.py add a JS file into the out put directory.

**Comment by Paebbels** [comment]
Hi. I'm back with a few questions :).

I would like to ask, how you computed the branch coverage value in percent compare the the statement coverage in %.

I used some time to update and fine tune my Sphinx directive `report:code-coverage`.  
Comparing results with your HTML report and also comparing with the raw data from `coverage.json` shows a branch coverage of circa 31%.

```json
{
	"meta": {
		"format": 3,
		"version": "7.8.0",
		"timestamp": "2025-04-16T09:20:36.774625",
		"branch_coverage": true,
		"show_contexts": false
	},
	"files": {

        },
	"totals": {
		"covered_lines": 500,
		"num_statements": 1341,
		"percent_covered": 30.91579594345421,
		"percent_covered_display": "31",
		"missing_lines": 841,
		"excluded_lines": 7,
		"num_branches": 286,
		"num_partial_branches": 3,
		"covered_branches": 3,
		"missing_branches": 283
	}
}

```

(1) This is how I computed the statement coverage:  
$statementCoverage = coveredStatements / totalStatements$  
Here I get 37% for statement coverage.  
I assume the difference is due to enabled branch coverage, because $branchCoverage <= statementCoverage$.

(2) What's still a question to me: Should it include the excluded statements too?  
I'm also unsure why you named the total number of statements `num_statements`, while the coverage uses `lines` in its names. Technically, there is a difference between statements coverage (each individual statement) and line coverage, because there can be multiple statements per line.

(3) How to compute the branchCoverage value per Python module. You can see my results in the table below. I'm completely off :)

I'm currently creating a reduced datamodel from your JSON file. It doesn't track statements and branches.  
(4) Or do I need to trace execution paths and count statements on these branches to get to the value?

----------------
Current state of the my extension looks like this.

Provided directives:
```rst
.. report:code-coverage::
   :packageid: src
```
and
```rst
.. report:code-coverage-legend::
   :packageid: src
   :style: vertical-table
```

<img width="969" alt="Image" src="https://github.com/user-attachments/assets/c2e7baa0-3fcd-4dcf-9c09-f0492598e61a" />
(The side-by-side view is created by Sphinx-Design and a layout grid.)

Example page: https://pytooling.github.io/sphinx-reports/CodeCoverage.html

Similarly, you'll find documentation coverage as a report page and unit tests as a page.

**Comment by nedbat** [comment]
@Paebbels I haven't read everything closely, but I would hope that the JSON report from coverage would have all of the data you need, and you wouldn't have to compute anything.  Maybe I've forgotten a gap in the report?

**Comment by Paebbels** [comment]
Currently the JSON has one coverage entry (`percent_covered`) per object. It's unclear if this is branch or statement coverage. I assume branch coverage, because it's currently active in the example JSON (see above).
Maybe it could help to provide both numbers.

Besides having all data available in the JSON file itself, all my data models I'm writing are able to verify, recompute or sanity check the presented data. In many cases I was able to report bugs to the original authors.

I'm not hoping to find any bugs in your code, but is a priniple of mine to be able to verify data files for consistency and manipulations. Especially, as coverage JSON files might have been merged.

So what would be needed to compute that number?
